sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -176,7 +179,9 @@ case class AdaptiveSparkPlanExec(
                 }
               }(AdaptiveSparkPlanExec.executionContext)
             } catch {
-              case e: Throwable => events.offer(StageFailure(stage, e))
+              case e: Throwable =>
+                earlyFailedStage = Some(stage.id)
```

Comment:
Can we call `cleanUpAndThrowException` here instead?

Reply:
We'll stop materializing following stages once we catch exception here and call `cleanUpAndThrowException` soon. So I think it may doesn't make big difference if you mean to call it earlier.

Also, we need to do extra stuff for the exception below.

Reply:
Does it still count as `Failed to materialize query stage` when initializing lazy val fails? If it counts then it seems better to send an event here.

Reply:
What extra stuff do we need to do? The event queue does not matter once you catch an active exception.
My idea was in the catch block, we just do:
```
errors.prepend(new SparkException(...)) // this should be the primary exception
cleanUpAndThrowException(errors, Some(stage.id))
```
And we can probably make this exception message slightly different from other stage failure messages.